src/Development/Shake/Internal/Rules/File.hs

```diff
@@ -313,6 +335,30 @@ fileForward act = addUserRule $ FileRule $ fmap ModeForward . act
 need :: [FilePath] -> Action ()
 need = void . apply_ fileNameFromString
 
+
+-- | Like 'need' but returns a list of rebuild dependencies this build.
+--
+--   The following example writes a list of changed dependencies to a file as its action.
+--
+-- @
+-- \"target\" '%>' \\out -> do
+--       let sourceList = [\"source1\", \"source2\"]
+--       rebuildList <- 'needHasChanged' sourceList
+--       'Development.Shake.writeFileLines' out rebuildList
+-- @
+--
+--   This function can be used to alter the action depending on which dependency needed
+--   to be rebuild.
+--
+--   Notice that a rule can be run even if no dependency has changed and needHasChanged
+--   will return an empty list on every call then. In this case a target got inconsistent
+--   and you should recreate it, without reusing any part of the target from the last build,
+--   as the target could have been edited or even deleted.
+needHasChanged :: [FilePath] -> Action [FilePath]
+needHasChanged paths = do
+    changed <- fmap hasChanged <$> apply_ fileNameFromString paths
+    return $ fmap fst $ filter snd $ zip paths changed
```

Comment:
Both those `fmap`'s are secret `map`s, so better to be explicit. However, I'd probably do:

```
res <- apply_ fileNameFromString paths
return [a | (a,b) <- zip paths res, hasChanged b]
```

That seems cleaner.

Reply:
That is a lot clearer.